### ISSUE TITLE
Readme day

### ISSUE THREAD

**Issue by des-des** [foundersandcoders/master-reference#17]
Right now I think there is general agreement that the current readme day does not work as it should.

**Comment by iteles** [comment]
The consensus in the room was that there is value in it, but it definitely needs to be restructured and its' aims need to be redefined.

**Comment by nelsonic** [comment]
The _purpose_ of writing READMEs needs to be _unambiguous_ and should not be _artificially confined_ to a _single_ day; writing READMEs should be an _integral_ part of getting _work_ done. You are documenting _`while`_ solving challenges and _systematically communicating_ with your team (_and/or "future self"_).

# tl;dr

## _Why_?

> There's _nothing_ more frustrating than joining (_or inheriting_) a technical project without docs! 😢 

Having to _forensically_ dig through a mountain of un-documented/commented code wastes time and frustrates new people and is the _biggest_ reason projects `fail`. 💩 

> Read: http://tom.preston-werner.com/2010/08/23/readme-driven-development.html

## _What_?

The original idea behind encouraging people to write READMEs was to _document_ the work they are doing _`before`_ (_or `while`_) building the project and encapsulate project-related knowledge so that it is not _lost_ and thus _always_ have the project in a "_reproducible_" and _beginner-friendly_ state.  

> Learn by _Teaching_: https://en.wikipedia.org/wiki/Learning_by_teaching

I don't know _when_ or _how_ it was decided to allocate a _day_ to writing READMEs and I _totally understand_ that people can _easily_ get frustrated by this rather vague idea/instruction. 😞 
Writing the README for your _project_ needs to be done _throughout_ the project not just on _one_ isolated day in pursuit of "_academic_" knowledge. If people (_in the cohort_) are having to create knowledge for tools/technologies _from scratch_ during the course their time is being _wasted_. 🔥 

> There is already _too much_ to learn in the **first 10 weeks** to spend time going on frustrating tangents, but discovering new techniques/frameworks once you have a _solid_ base is _essential_ for "[_continuous professional development_](https://en.wikipedia.org/wiki/Professional_development)".

## How?

> For the first **10 weeks** of the F&C there is at least a _beginner_-level README for every subject/tool/technology: https://github.com/search?q=org%3Adwyl+learn 
in _many_ cases the `dwyl` resource is the most popular tutorial on the `internet`... 
it's not an _accident/coincidence_...  we have _systematically_ written the READMEs to _condense_ our knowledge, ***save people time*** ⌛️ and "[***out-teach***](http://headrush.typepad.com/)" the "_competition" ...
> Watch: http://ecorner.stanford.edu/videos/2361/Out-Teach-Your-Competition to understand _Why_ we have so many learning-focussed repos and why we will continue to _expand_ our knowledge. https://youtu.be/XDGHxO6N3Ms?t=18m31s

Some _previous_ F&C cohorts have used `dwyl` READMEs to learn principals/techniques/tools and have _contributed_ back by _extending_ the knowledge (_either with **clarifying questions**, pull requests_) and/or simply _starring_ the repo(s)... ⭐️ 
But I can count on _one hand_ how many **FAC9** people have asked a _clarifying_ question on one of the DWYL `learn-...` repos, so I have _zero visibility_ of what is "_confusing_" people... 😕  ❓
_Why_ is this? Are people "_afraid_" of asking questions? Or did we just not make it _clear_  that asking questions is _essential_ to learning...? 🤔  
Did someone tell **FAC9** people _**NOT**_ to use the `dwyl` resources...? 
Is there a _masochistic_ preference to _waste_ time searching even when a resource already exists?! 

> The purpose of F&C is to _empower_ people with the skills they need to leverage technology to solve _real_ challenges and make the world a better place.
> See: https://github.com/foundersandcoders/oursite/issues/2 and https://github.com/foundersandcoders/business/issues/40

**Comment by bradreeder** [comment]
This was the second highest priority problem as put to the vote by FAC9 #18. Given that FAC8 also had a lot of issues with readme day I think we should treat this as a high priority issue to be resolved by the time we get back to curriculum planning post Christmas.

Their sticky notes included:

* Keeping presentations to a strict time limit.
* Walkthrough tutorials instead of readmes on some of the hardest topics.
* Extending project time and making scheduled research time apart of the project.
* If they need to know something for a project, don't assign it to a readme topic, include it in a workshop instead.

**Comment by bradreeder** [comment]
Readme day was something I railed against most S-G-C's in my F&C. It's simply inconceivable that you can create a useful resource on a topic, that everyone [anyone?] in the group will learn something substantial from, from no prior knowledge within 4 hours. It's a microcosm of my frustration with the way curriculum planning has been approached until now.

Realistically readme afternoons are a means of cramming topics into a week that mentors didn't have time to do prepare proper workshops for. I also understand in theory that readme day is meant to teach you soft skills:

* Researching unfamiliar topics independently.
* Practice at documentation.
* Practice at presentation.

I think @nelsonic has already said this well, but the issue is that the practice of readme afternoon's is detached from the actual context in which you would practice these things as a developer. 

Documentation and research into unknowns should be integrated into the practice of writing code, not be seen as separate, and delivering a well-prepared presentation should be a part of a team's weekly project demo.

My suggestion going forward would be something like this:

* Drop readme afternoons on Tuesday's.
* Given that Monday's now have 2 hours of 5% time let's have two more hours of workshops on a Tuesday to make up for it.
* Let's use the remaining time on a Tuesday to start the project's early (perhaps as a planning afternoon) and discuss ways that allocated research time and documentation can be incorporated into it. 
* Let's start preparing a schematic for what weekly project presentations should include and give students allocated time to prepare them well.

**Comment by des-des** [comment]
@nelsonic @bradreeder 

Gonna try and pull a vague list of requirements out here.
 1. Students need to learn about documenting their learning, so that they can properly spread knowledge and document their projects.
 2. Before projects start, students need to understand how to use the required technologies to achieve their project aims, while also being able to test in the tight timescales allocated. If we can put students in better shape to complete their projects _and_ give them a better understanding of what tests they can write, they may be more successful writing properly tested code.
 3. (_optional_) students should learn how to research then use an unknown technology quickly and effectively.

@bradreeder It is on my last point that I feel _more_ workshop time may fall down.. The other thing here is how much teaching time we require from mentors, we want to be a little careful of setting things up in a way that will fail if we do not have lots of experienced devs hanging around willing to give up their time..

**Comment by bradreeder** [comment]
@des-des on the last point I'd clarify that we used to have 2 more hours of workshops on Mondays that have been replaced with business development. This means we'd only have 2 workshops per week with the current schedule #27 , Monday morning + afternoon. I'm only proposing we move these 2 hours to Tuesday, not add more.

**Comment by des-des** [comment]
Yes, having had a look at the proposed schedule this sounds good. We have 2hours of morning challenges + 4/5 hours of workshops. I think this is good.

I still like the idea of some sort of spike / independent learning on tuesday before the projects start..

**Comment by Cleop** [comment]
My own personal view on READMEs:

- I like having the chance to work more quietly/independently from time to time, it helps break things up a bit.
- I think it can be beneficial to force us to learn new things that otherwise we've perhaps never heard of (this could also happen in the format of workshops, however can we cover as much in as little time?)
- I think presentations and their use varies considerably according to the amount of time given to do them (not too long) and the clarity of instructions for how the research should be conducted/presented in the first place. As a reflection of this, I did find some README sessions extremely useful, others not at all.
- I agree with the fact that given our time constraints I think that the Tuesday sessions are perhaps misleadingly called READMEs because I don't think what we create in that time/ for the purpose of presentation actually resembles a regular README.

I'm not opposed to getting rid of them given the time constraints we have from business dev etc. and how they have the potential to be ineffective. However, it'd be great if we could not to lose some of the subjects they covered altogether and if there's any other way we can think of to give us a bit of respite from team/ board walk-through working to refresh our brains.

**Comment by sofer** [comment]
A bit of historical perspective: README day was originally called *"jigsaw"* day after [the jigsaw classroom](https://www.jigsaw.org/) developed by Elliot Aronson. One student from each project team joined a four-person jigsaw group that spent time together learning some piece of information critical to the delivery of the week's project. 

*"Jigsaw"* day became *"Research"* (and then *"README"*) day because at some point around FAC6 or FAC7 we decided that paired working was more useful than teams of four.

I agree wholeheartedly with @Cleop that both the scope of the research and the length of presentations should be kept short. Some careful guidance and topic selection is needed.

**Comment by bradreeder** [comment]
@sofer @Cleop  I think we could reduce the scope of readme afternoons to 3 hours if we introduced the projects Tuesday afternoon and said something like 'in order to complete these projects you will need to research into the following unknowns', then jigsaw the group to research into those unknowns? 

Rather than present to the room, which takes time, the following morning each team could have a stand-up meeting, in agile fashion, in which each person reports on their progress, any obstructions the group might experience with the topic, and talks through a practical example, or a spike like @des-des says, with the group? 

This could lead neatly into their planning for the project as they'll be in a better position to estimate how much time each task will take. Rather than spending time generating readmes they could document the results of their research in their project's readme?

**Comment by des-des** [comment]
@bradreeder I think this sounds good. I think it would be good to build small exercises for this. In the situation where we are moving from vague readme topic to a focused spike, we might need to more direction than we have had before.

I think if we need to move away from the idea of readme creation, we need to find a way to capture the importance of readmes in the curriculum.

Picking up on one of @Cleop 's point, I would have personally liked some time for independent learning during the course. I think that sometimes people end up wanting a little space. I think it might be worth thinking about that a bit..

**Comment by bradreeder** [comment]
@des-des Yeah, definitely on all those points. I felt the introvert exhaustion muchly. :P

1) Independent time: Say after lunch we did a quick whiteboard q/a & introduced the project, the rest of the afternoon could be each team member's chance to research a part of what the project involves independently before working with the team the next morning?

2) Readmes: If the focus did become on a spike, on communicating to the rest of your team unfamiliar code, maybe what we could teach through that exercise is how to document your code? Teams could focus on making their project readme as informative & well documented as possible, & the value of that could be shown through how much easier it makes it for other groups to code review them?

**Comment by msachi** [comment]
The main problem I see with the readme day is that, because of the small amount of time, the large scope of topics and the fact that most students are starting from scratch, students _do not understand_ the material that they are presenting. It often results in copy-pasting large quantities of text and then reading from slides, which obviously isn't helpful to anyone listening.

Ensuring that presentations are kept short and that students only write down the _summary_ of the material (i.e. how they understand it) helps, but if the readme day was to survive in its current format, it would be equally important to 1) reduce the scope of topics, 2) make them more practical and less abstract, and 3) direct students to specific technologies or resources. This could mean asking students to implement hapi-auth-basic in a mini-project rather than asking them to 'cover the topic of authentication'. The main aim of the readme day then would be to prepare students for the projects.

I think this could work in the current format, but the format that @bradreeder proposes also sounds great and it would be worth trying out. I also like the idea of putting more stress of project readmes, and ensuring that students pick up the 'readme creation' skills there. My only concern is with the independent time, because I do value it and I think it's useful for independent learning, but I think that doing these 'spikes' in a team would produce a better result than individually. (This I think is another problem with the readme day as it is - each person working on 'one of the 4 questions' often results in a fragmented result, and in students having little awareness of the rest of the team members work.)
